WordPress/src/main/java/org/wordpress/android/ui/JetpackConnectionWebViewClient.java

```diff
@@ -0,0 +1,137 @@
+        view.postUrl(authenticationURL, postData.getBytes());
+    }
+
+    private String extractRedirect(String stringUrl) throws UnsupportedEncodingException {
+        int from = stringUrl.indexOf(REDIRECT_PARAMETER) + REDIRECT_PARAMETER.length();
+        int to = stringUrl.indexOf("&", from);
+        String redirectUrl;
+        if (to > from) {
+            redirectUrl = stringUrl.substring(from, to);
+        } else {
+            redirectUrl = stringUrl.substring(from);
+        }
+        if (redirectUrl.startsWith(JETPACK_PATH)) {
+            redirectUrl = WORDPRESS_COM_PREFIX + redirectUrl;
+        }
+        return URLDecoder.decode(redirectUrl, WPWebViewActivity.ENCODING_UTF8);
+    }
+
+    @Override
+    public boolean shouldOverrideUrlLoading(WebView view, String stringUrl) {
+        try {
+            final Uri url = Uri.parse(stringUrl);
+            final SiteModel site = mSiteStore.getSiteByLocalId(AppPrefs.getSelectedSite());
+            final String loadedHost = url.getHost();
+            final String loadedPath = url.getPath();
+            final String currentSiteHost = Uri.parse(site.getUrl()).getHost();
+            if (loadedHost.equals(currentSiteHost)
+                    && loadedPath.equals(LOGIN_PATH)
+                    && stringUrl.contains(REDIRECT_PARAMETER)) {
+                redirectPage = extractRedirect(stringUrl);
+                loginToWPCom(view, site);
+                return true;
+            } else if (loadedHost.equals(currentSiteHost)
+                    && loadedPath.equals(ADMIN_PATH)
+                    && redirectPage != null) {
+                view.loadUrl(redirectPage);
+                redirectPage = null;
+                return true;
+            } else if (loadedHost.equals(WORDPRESS_COM_HOST)
+                    && loadedPath.equals(LOG_IN_PATH)
+                    && stringUrl.contains(REDIRECT_PARAMETER)) {
+                redirectPage = extractRedirect(stringUrl);
+                Intent loginIntent = new Intent(activity, LoginActivity.class);
+                LoginMode.JETPACK_STATS.putInto(loginIntent);
+                activity.startActivityForResult(loginIntent, REQUEST_CODE);
+                return true;
+            } else if (loadedHost.equals(JETPACK_DEEPLINK_URI.getHost())
+                    && url.getScheme().equals(JETPACK_DEEPLINK_URI.getScheme())) {
+                Intent intent = new Intent(Intent.ACTION_VIEW);
+                intent.setData(url);
+                activity.startActivity(intent);
+                activity.finish();
+                flowFinished = true;
+                AnalyticsTracker.track(AnalyticsTracker.Stat.STATS_SELECTED_INSTALL_JETPACK);
+                return true;
+            }
+        } catch (UnsupportedEncodingException e) {
+            AppLog.e(AppLog.T.API, "Unexpected URL encoding in Jetpack connection flow.", e);
+            e.printStackTrace();
```

Comment:
`AppLog.e` already logs the stack trace, so `e.printStackTrace()` isn't necessary here.